synapse/lib/oauth.py

```diff
@@ -185,39 +158,87 @@ async def _oauthRefreshLoop(self):
         formdata.add_field('scope', providerconf['scope'])
         formdata.add_field('redirect_uri', providerconf['redirect_uri'])
         formdata.add_field('code', authcode)
         if code_verifier is not None:
             formdata.add_field('code_verifier', code_verifier)
 
-        auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
         return await self._fetchOAuthToken(token_uri, auth, formdata, ssl_verify=ssl_verify)
 
-    async def _refreshOAuthAccessToken(self, providerconf, clientconf):
+    async def _refreshOAuthAccessToken(self, providerconf, clientconf, useriden):
+        formdata = aiohttp.FormData()
         token_uri = providerconf['token_uri']
         ssl_verify = providerconf['ssl_verify']
         refresh_token = clientconf['refresh_token']
+        client_assertion = providerconf.get('client_assertion')
 
-        formdata = aiohttp.FormData()
         formdata.add_field('grant_type', 'refresh_token')
         formdata.add_field('refresh_token', refresh_token)
 
-        auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
+        if client_assertion:
+            auth = None
+            ok, info = await self._getClientAssertionData(providerconf, formdata, useriden)
+            if not ok:
+                return ok, info
+        else:
+            auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
+
         ok, data = await self._fetchOAuthToken(token_uri, auth, formdata, ssl_verify=ssl_verify, retries=3)
         if ok and not data.get('refresh_token'):
             # if a refresh_token is not provided in the response persist the existing token
             data['refresh_token'] = refresh_token
 
         return ok, data
 
+    async def _getClientAssertionData(self, providerconf, formdata, useriden):
+
+        client_id = providerconf['client_id']
+        client_assertion = providerconf['client_assertion']
+
+        if (info := client_assertion.get('cortex:callstorm')):
+            opts = {
+                'view': info['view'],
+                'vars': info.get('vars', {}),
+                'user': useriden,
+            }
+            try:
+                ok, info = await self.callStorm(info['query'], opts=opts)
+            except Exception as e:
+                return False, {'error': f'Error executing callStorm: {e}'}
+            if not ok:
+                return ok, info
+            assertion = info.get('token')
+        elif client_assertion.get('msft:azure:workloadidentity'):
+            ok, info = _getAzureTokenFile()
+            if not ok:
+                return ok, info
+            assertion = info.get('token')
+        else:  # pragma: no cover
+            return False, {'error': f'unknown client_assertions data: {client_assertion=}'}
+
+        formdata.add_field('client_assertion_type', 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer')
```

Comment:
It seems a bit odd to use a retn and still directly mutate the formdata and return `True, {}`. Just a little weird.

Reply:
was using the retn more like a c error code. we could definitely pass back useful objects instead, especially since i'm looking to rework the "configure auth for a request" anyhow.

Reply:
This has been rewritten.